src/System.Memory/src/System/SpanHelpers.byte.cs

```diff
@@ -95,82 +105,218 @@ public static int IndexOf(ref byte searchSpace, byte value, int length)
+                    }
+
+                    index += LocateFirstFoundByte(flaggedMatches);
+                    goto exitFixed;
+                }
+
+                for (; index < length; index++)
+                {
+                    if (*(haystack + index) == value)
+                    {
+                        goto exitFixed;
+                    }
+                }
+                // No Matches
+                index = -1;
+                // Don't goto out of fixed block
+                exitFixed:;
+            }
+            exit:
+            return index;
+        }
 
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int LocateFirstFoundByte(Vector<byte> match)
+        {
+            var vector64 = Vector.AsVectorUInt64(match);
+            ulong candidate = 0;
+            var i = 0;
+            // Pattern unrolled by jit https://github.com/dotnet/coreclr/pull/8001
+            for (; i < Vector<ulong>.Count; i++)
+            {
+                candidate = vector64[i];
+                if (candidate == 0) continue;
+                break;
+            }
 
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
+            // Single LEA instruction with jitted const (using function result)
+            return i * 8 + LocateFirstFoundByte(candidate);
+        }
 
-                length -= 4;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int LocateFirstFoundByte(ulong match)
+        {
+            unchecked
+            {
+                // Flag least significant power of two bit
+                var powerOfTwoFlag = match ^ (match - 1);
+                // Shift all powers of two into the high byte and extract
+                return (int)((powerOfTwoFlag * xorPowerOfTwoToHighByte) >> 57);
             }
+        }
 
-            while (length > 0)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong SetLowBitsForByteMatch(ulong potentialMatch, byte search)
+        {
+            unchecked
             {
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-                length--;
+                var flaggedValue = potentialMatch ^ (byteBroadcastToUlong * search);
+                return (
+                        (flaggedValue - byteBroadcastToUlong) &
+                        ~(flaggedValue) &
+                        filterByteHighBitsInUlong
+                       ) >> 7;
             }
+        }
 
-            return true;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector<byte> GetVector(byte vectorByte)
+        {
+#if !NETCOREAPP1_2
```

Comment:
How will that work with System.Memory package? i.e. do we now need three different dlls in the package: one for below 1.2, one for 1.2, and one for 2.0?

cc: @ahsonkhan